File: LeetCode/2962/Solution.java
class Solution {
    public long countSubarrays(int[] nums, int k) {
        int n = nums.length;

        int max = 0;
        for(int num : nums) {
            max = Math.max(max, num);
        }

        int start = 0;
        int end = 0;
        int[] count = new int[1000001];
        long result = 0;
        while(end < n) {
            count[nums[end]]++;
            while(count[max] >= k) {
                count[nums[start]]--;
                result += n-end;
                start++;
            }
            end++;
        }
        return result;
    }
}

File: LeetCode/2962/Solution.kt
class Solution {
    fun countSubarrays(nums: IntArray, k: Int): Long {
        val counter = mutableMapOf<Int, Int>()

        var result = 0L
        var left = 0
        var right = 0
        var maxValue = nums.max()

        while(right < nums.size) {
            val v = nums[right]
            counter[v] = counter.getOrDefault(v, 0)+1
            while(counter.getOrDefault(maxValue, 0) >= k) {
                val vv = nums[left]
                counter[vv] = counter.getOrDefault(vv, 0)-1
                result += nums.size-right
                left++
            }
            right++
        }
        return result
    }
}
